src-built-in/components/linker/src/linker.tsx

```diff
@@ -0,0 +1,75 @@
+/*!
+* Copyright 2017 by ChartIQ, Inc.
+* All rights reserved.
+*/
+import * as React from "react";
+import { useEffect } from "react";
+import { Provider, useSelector, useDispatch } from 'react-redux';
+import * as ReactDOM from "react-dom";
+
+import "./css/linkerWindow.css";
+import "../../../../assets/css/font-finance.css";
+import "../../../../assets/css/finsemble.css";
+import { RootState } from '../../UIAPI/types';
+import store from '../../UIAPI/store';
+import * as linkerActions from "../../UIAPI/actions/linkerActions";
+
+declare var FSBL: any;
+declare global {
+    interface Window {
+        FSBL: any
+    }
+};
+
+const LinkerRefactored = () => {
+	const linker = useSelector((state: RootState) => state.linker);
+	const dispatch = useDispatch();
+	
+	const toggleChannel = (linkerIndex: number) => {
+		dispatch(linkerActions.toggleChannel(linkerIndex));
+	};
+
+    useEffect(() => {
+		dispatch(linkerActions.init());
+        return () => {
+            dispatch(linkerActions.cleanUp());
+        }
+	}, []);
+
+	const allChannels = Object.values(linker.channels);
+    const channelsElements = allChannels.map(({color, name, active, id}) => {
+        const groupClass = `linkerGroup ${color}`;
+        const style = {
+            backgroundColor: color,
+            border: `1px solid ${color}`
+		}
+        return (
+            <div key={id} className="channel-wrapper" onClick={() => toggleChannel(id)}>
+                {linker.isAccessibleLinker ? <div className="channel-label">{name}</div> : null}
+				<div className={groupClass} style={style}>
+					{active ? <i className="active-linker-group ff-check-mark"></i> : null}
+				</div>
+            </div>
+        );
+    });
+
+    return (
+        <div className="linkerContainer">
+            {channelsElements}
+        </div>
+    );
+}
+
+if (window.FSBL && FSBL.addEventListener) {
```

Comment:
@stephen-ciq Would like to see this finsemble boilerplate moved inside the Provider.

Reply:
@thorsent I apologize for just now getting to this.

in a short convo with Zona about how to fulfill your request, the best thing I could come up with was this pseudocode.  it puts the logic of "is fsbl ready" in the store, which is provided by the provider, and no component other than the FSBLReadyGuard has to worry about it.  please provide feedback as to whether this meets your expectations when you get a chance =)

```
function addFSBLReadyListener ()
{
  if (window.FSBL && FSBL.addEventListener) {
	FSBL.addEventListener("onReady", FSBLReady);
  } else {
	window.addEventListener("FSBLReady", FSBLReady)
  }
}

function FSBLReducer(state, action)
{
switch (action.type)
case init:
  return loop(state, Cmd.run(addFSBLReadyListener, ...);
case FSBLReady:
  state.fsblIsReady = action.payload.fsblIsReady;
...
}

function FSBLReadyWrapper(props.children) {
    fsblReadySelector = useSelector(fsblIsReady );
    if (fsblReadySelector())
      // return props.children
    else
     // return "Loading" or <div /> or whatever
}
function FSBLReady() {
	ReactDOM.render(
	<Provider store={store}>
            <FSBLReadyGuard >
    		    <LinkerRefactored />
            </FSBLReadyGuard >
  	</Provider>,
   	document.getElementById("main")
   	);
}
```

Reply:
@zona-ciq, @mesner mentioned to me that there are library facilities for async initialization that would help make this pseudocode a reality/clean it up and make it clearer.  I'd suggest looking at those/chatting with him.